pkg/fqdn/service/service_test.go

```diff
@@ -0,0 +1,244 @@
+	testidentity "github.com/cilium/cilium/pkg/testutils/identity"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+func TestFQDNDataServer(t *testing.T) {
+	// testutils.PrivilegedTest(t)
+
+	test := map[string]struct {
+		port                     int
+		enableL7Proxy            bool
+		enableStandaloneDNSProxy bool
+		serverPort               int
+		err                      error
+	}{
+		"Successfully running the server": {
+			port:                     1234,
+			enableL7Proxy:            true,
+			enableStandaloneDNSProxy: true,
+			serverPort:               40045,
+			err:                      nil,
+		},
+		"Failure on running the server due to invalid server port": {
+			port:                     1234,
+			serverPort:               0,
+			enableStandaloneDNSProxy: true,
+			enableL7Proxy:            true,
+			err:                      errors.New("listen tcp: address -1: invalid port"),
+		},
+	}
+
+	for scenario, tt := range test {
+		t.Run(scenario, func(t *testing.T) {
+
+			em := endpointmanager.New(&dummyEpSyncher{}, nil, nil, nil, nil)
+			pr := policy.NewPolicyRepository(hivetest.Logger(t), nil, nil, nil, nil, api.NewPolicyMetricsNoop())
+			ipc := ipcache.NewIPCache(&ipcache.Configuration{
+				Context:           context.TODO(),
+				IdentityAllocator: testidentity.NewMockIdentityAllocator(nil),
+				PolicyHandler:     pr.GetSelectorCache(),
+				DatapathHandler:   em,
+			})
+			nm := namemanager.New(namemanager.ManagerParams{
+				Config: namemanager.NameManagerConfig{
+					MinTTL:            1,
+					DNSProxyLockCount: defaults.DNSProxyLockCount,
+					StateDir:          defaults.StateDir,
+				},
+				IPCache: ipc,
+			})
+
+			hive := hive.New(
+				Cell,
+				cell.Provide(func() *ipcache.IPCache {
+					return ipc
+				}),
+				cell.Provide(func() endpointmanager.EndpointManager {
+					return em
+				}),
+				cell.Provide(func() namemanager.NameManager { return nm }),
+				cell.Provide(func(lc cell.Lifecycle, p health.Provider, jr job.Registry) job.Group {
+					h := p.ForModule(cell.FullModuleID{"test"})
+					jg := jr.NewGroup(h)
+					lc.Append(jg)
+					return jg
+				}),
+				cell.Provide(func(lc cell.Lifecycle) messagehandler.DNSRequestHandler {
+					return messagehandler.NewDNSRequestHandler(
+						messagehandler.DNSRequestHandlerParams{
+							Lifecycle:         lc,
+							Logger:            hivetest.Logger(t),
+							NameManager:       nm,
+							ProxyInstance:     nil,
+							ProxyAccessLogger: nil,
+						})
+				}),
+				cell.Invoke(newServer),
+			)
+
+			flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+			hive.RegisterFlags(flags)
+			// Set the flags for the fqdn server
+			flags.Set("enable-standalone-dns-proxy", fmt.Sprintf("%t", tt.enableStandaloneDNSProxy))
+			flags.Set("to-fqdns-server-port", fmt.Sprintf("%d", tt.serverPort))
+
+			// Set the flags for the l7 proxy(from cilium agent)
+			option.Config.EnableL7Proxy = tt.enableL7Proxy
```

Comment:
I think it'd be worth checking which things are accessing these fields and seeing if we could instead (as a first step) set these via a provided `*option.DaemonConfig` instead of going through the global, e.g. `cell.Provide(func() *option.DaemonConfig { return &option.DaemonConfig{ ... })`.

Using the global variable is problematic as setting these here will affect all other tests in the same package. Though probably not an issue here, but it's usually good to always avoid using `option.Config` directly and instead try to go via the `*option.DaemonConfig` parameter. Of course best would be to not have the `option.Config` at all, but that's still long way off...

Reply:
I updated to use `*option.DaemonConfig`, let me know if this looks good.